src/IECoreRenderMan/MaterialCache.cpp

```diff
@@ -0,0 +1,110 @@
+//////////////////////////////////////////////////////////////////////////
+//
+//  Copyright (c) 2024, Cinesite VFX Ltd. All rights reserved.
+//
+//  Redistribution and use in source and binary forms, with or without
+//  modification, are permitted provided that the following conditions are
+//  met:
+//
+//      * Redistributions of source code must retain the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "MaterialCache.h"
+
+#include "ShaderNetworkAlgo.h"
+
+using namespace std;
+using namespace IECore;
+using namespace IECoreScene;
+using namespace IECoreRenderMan;
+
+MaterialCache::MaterialCache( const Session *session )
+	:	m_session( session )
+{
+}
+
+ConstMaterialPtr MaterialCache::getMaterial( const IECoreScene::ShaderNetwork *network )
```

Comment:
I found the slight differences between this and `getDisplacement()` a little confusing, maybe I'm missing the subtleties but shouldn't they basically do the same thing with different caches?

Reply:
Indeed. There were more differences at one point in development, with the MaterialCache implementing a fallback to a facing ratio shader for null materials. I moved that to Attributes, but in doing so left behind some unnecessary differences with the displacement code. I've brought the two into alignment in https://github.com/GafferHQ/gaffer/pull/6271/commits/12d0a237196b15e92f3f84f83a2978a992ef8ba0.